Public/Src/Cache/ContentStore/Library/Service/Grpc/BufferedReadStream.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace BuildXL.Cache.ContentStore.Service.Grpc
+{
+
+    // This class is required for decompressing data streamed via grpc
+    // because the compression api insists on reading from a stream and
+    // the grpc api exposes a read-one-chunk method instead of a stream.
+
+    internal class BufferedReadStream : Stream
+    {
+        public BufferedReadStream(Func<Task<byte[]>> reader)
+        {
+            if (reader is null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+            _reader = reader;
+        }
+
+        private readonly Func<Task<byte[]>> _reader;

```

Comment:
nit: we don't have specific guidelines, but in most cases we put the fields on top.